src/lexer.rs

```diff
@@ -15,17 +16,17 @@ type CharAndLocation = (char, Location);
 //      affected by the removal of newlines in phase 6 of compilation.
 //
 // It expects that phases 1 to 3 are already done and that valid utf8 is passed in.
-#[derive(Clone, Copy)]
+#[derive(Clone)]
 pub struct CharsAndLocation<'a> {
-    input: &'a str,
+    chars: Peekable<Chars<'a>>,
```

Comment:
I'm trying to understand why this change was needed because it makes `CharsAndLocations` no longer Copyable trivially. But I can't see what changes in the logic to require it.

Reply:
Isn't rebuilding the iterator free? After checking it isn't 100% free. It does:

```
    pub fn chars(&self) -> Chars<'_> {
        Chars { iter: self.as_bytes().iter() }
    }
```

Which as far as I understand just takes the slice's data as (ptr, len) and puts it in a struct that's (ptr_start, ptr_end). Everything is tagged as inline so the only thing that happens is an addition. So we're trading an addition for a slightly bigger copy + addition of `clone()` calls. I have a very slight preference for the code as it was previously but don't really mind either way.

Reply:
I'm not sure, I was checking godbolt and for debug it's clear that there's a performance hit but in release there isn't so I think I'll revert it

Reply:
I just double checked and forgot to save the `str` back when testing in godbolt that's why the previous version was faster (I simplified the examples a bit to remove some noise from the de-compilation) and now even with optimizations the version that doesn't rebuild the iter produces a better looking assembly.

No rebuilding version: https://godbolt.org/z/aT8Wd9Yb8

Rebuilding version: https://godbolt.org/z/651aY7snE

Reply:
Ack, let's keep it as in the current version of the PR then.